src/category_theory/limits/shapes/diagonal.lean

```diff
@@ -0,0 +1,369 @@
+/-
+Copyright (c) 2022 Andrew Yang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Andrew Yang
+-/
+import category_theory.limits.shapes.pullbacks
+import category_theory.limits.shapes.kernel_pair
+import category_theory.limits.shapes.comm_sq
+
+/-!
+# The diagonal object of a morphism.
+
+We provide various API and isomorphisms considering the diagonal object `Δ_{Y/X} := pullback f f`
+of a morphism `f : X ⟶ Y`.
+
+-/
+
+open category_theory
+
+noncomputable theory
+
+namespace category_theory.limits
+
+variables {C : Type*} [category C] {X Y Z : C}
+
+namespace pullback
+
+section diagonal
+
+variables (f : X ⟶ Y) [has_pullback f f]
+
+/-- The diagonal object of a morphism `f : X ⟶ Y` is `Δ_{X/Y} := pullback f f`. -/
+abbreviation diagonal_obj : C := pullback f f
+
+/-- The diagonal morphism `X ⟶ Δ_{X/Y}` for a morphism `f : X ⟶ Y`. -/
+def diagonal : X ⟶ diagonal_obj f :=
+pullback.lift (𝟙 _) (𝟙 _) rfl
+
+@[simp, reassoc] lemma diagonal_fst : diagonal f ≫ pullback.fst = 𝟙 _ :=
+pullback.lift_fst _ _ _
+
+@[simp, reassoc] lemma diagonal_snd : diagonal f ≫ pullback.snd = 𝟙 _ :=
+pullback.lift_snd _ _ _
+
+instance : split_mono (diagonal f) :=
+⟨pullback.fst, diagonal_fst f⟩
+
+instance : split_epi (pullback.fst : pullback f f ⟶ X) :=
+⟨diagonal f, diagonal_fst f⟩
+
+instance : split_epi (pullback.snd : pullback f f ⟶ X) :=
+⟨diagonal f, diagonal_snd f⟩
+
+instance [mono f] : is_iso (diagonal f) :=
+begin
+  rw (is_iso.inv_eq_of_inv_hom_id (diagonal_fst f)).symm,
+  apply_instance
+end
+
+/-- The two projections `Δ_{X/Y} ⟶ X` forms a kernel pair for `f : X ⟶ Y`. -/
```

Comment:
```suggestion
/-- The two projections `Δ_{X/Y} ⟶ X` form a kernel pair for `f : X ⟶ Y`. -/
```

Reply:
You can check that [the version bors attempted to push](https://github.com/leanprover-community/mathlib/commit/63067d4c76074fe4fd8b3789753956773addd25b) is identical to [the diff until the penultimate commit](https://github.com/leanprover-community/mathlib/pull/15711/files/cc903c5fd38d5506a05ff9ae9ee634664c17dce0).
Or you may wait for Johan if you have any further doubts.

I accidentally rebased master instead of merging it.